finality-aleph/src/sync/substrate.rs

```diff
@@ -0,0 +1,46 @@
+use std::hash::{Hash, Hasher};
+
+use sp_runtime::traits::{CheckedSub, Header as SubstrateHeader, One, UniqueSaturatedInto};
+
+use crate::sync::{BlockIdentifier, Header};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BlockId<H: SubstrateHeader> {
+    hash: H::Hash,
+    number: H::Number,
+}
+
+impl<SH: SubstrateHeader> Hash for BlockId<SH> {
+    fn hash<H>(&self, state: &mut H)
+    where
+        H: Hasher,
+    {
+        self.hash.hash(state);
+        self.number.hash(state);
+    }
+}
+
+impl<H: SubstrateHeader> BlockIdentifier for BlockId<H> {
+    fn number(&self) -> u32 {
+        self.number.unique_saturated_into()
```

Comment:
If you specified `H` to `SubstrateHeader<Number = u32>` you would not have to use saturated into here. Also there exists `BlockNumber` in aleph-primitives, but is inconsistent with `BlockNumber` in bin (both are set separately to u32), so one of them could be used here instead of `u32`. If you decide to do that, it can be also used in BlockIdentifier trait.

Reply:
Oh, nice, fixed the first part, although the definitions became a bit noisy. I used the `BlockNumber` from `primitives`, but I haven't fixed the inconsistency...